File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#define min(a,b)  (((a) > (b)) ? (b) : (a))
#define max(a,b)  (((a) > (b)) ? (a) : (b))

int dfs(int cur, int n, int* visited, int** graph, int count){
    visited[cur] = 1;
    for (int i = 0; i < n; i++){
        if (graph[cur][i] && !visited[i]){
             int result = dfs(i, n, visited, graph, count+1);
             count = max(count, result);
        }
    }
    return count;
}

// wires_rows는 2차원 배열 wires의 행 길이, wires_cols는 2차원 배열 wires의 열 길이입니다.
int solution(int n, int** wires, size_t wires_rows, size_t wires_cols) {
    int answer = n-2;
    int **graph = (int**) calloc (n,sizeof(int*) );
    for(int i=0; i<n; i++){
        graph[i] = (int*) calloc (n, sizeof(int) );
    }
    for (int i = 0; i < wires_rows; i++){
        graph[wires[i][0]-1][wires[i][1]-1] = 1;
        graph[wires[i][1]-1][wires[i][0]-1] = 1;
    }

    for (int i = 0; i < wires_rows; i++){
        graph[wires[i][0]-1][wires[i][1]-1] = 0;
        graph[wires[i][1]-1][wires[i][0]-1] = 0;
        
        int* visited = (int*) calloc (n, sizeof(int));
        int con_num = dfs(0, n, visited, graph, 1);        
        answer = min(answer, abs(con_num-(n-con_num)));

        graph[wires[i][0]-1][wires[i][1]-1] = 1;
        graph[wires[i][1]-1][wires[i][0]-1] = 1;
    }

    for (int i = 0; i < n; i++) {
      free(graph[i]);
    }
    free(graph);
    
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    HashMap<Integer, ArrayList<Integer>> graph = new HashMap();

    int dfs(int start, int[] discon_wire, boolean[] visited, int count){
        visited[start] = true;
        ArrayList<Integer> nextList = graph.get(start);
        
        for (int next:nextList){
            if (Arrays.equals(discon_wire, new int[]{start, next}) || Arrays.equals(discon_wire,new int[]{next, start})) {
                continue;
            }
            if (!visited[next]){
                count = dfs(next, discon_wire, visited, count+1);
            }
        }
        return count;
    }
    
    public int solution(int n, int[][] wires) {
        int answer = n-2;
        
        for (int i = 0; i <n; i++){
            graph.put(i+1, new ArrayList<Integer>());
        }
        
        for (int[] wire:wires){
            ArrayList list1 = graph.get(wire[0]);
            list1.add(wire[1]);
            ArrayList list2 = graph.get(wire[1]);
            list2.add(wire[0]);
        }
        
        for (int[] wire:wires){
            boolean visited[] = new boolean[n+1];
            int result = dfs(1, wire, visited,1);
            answer = Math.min(answer, Math.abs(result-(n-result)));
        }
        return answer;
    }
}